### PR TITLE
Prosumer simulator

### PR DESCRIPTION
**Pull request by Hampusholmstrom**
Added prosumer buffer simulator model (net production / 10)

Takes care of under_production_buy and over_production_sell with an introduced schema variable "balance".

under_production_buy and over_production_sell cannot be "active" (over 0) at the same time. 

Small fixes of the simulator in case of unreal simulations (winds over 50 m/s etc.)

Removed server/sim_db_utils.js and sim/server_db_utils.js, were no longer needed because they do not hold the connection and were basically only dublicates.

### CODE REVIEW

File: server/api/routes.js [hunk 1]
```diff
@@ -1,10 +1,12 @@
...
 const router = express.Router()
-const server_db_utils = require('./../server_db_utils')
+const server_db_utils = require('../server_db_utils')
+const sim_db_utils = require('../../simulator/sim_db_utils')
 
 //CONSUMPTION
-//Update prosumer consumption by id
-router.put('/prosumer/:id/consumption', async (req, res) => {
+//Update prosumer.js consumption by id
+router.put('/prosumer.js/:id/consumption', async (req, res) => {
```
Comment (edvinnn) [inline server/api/routes.js]: `prosumer.js`... I guess it's a typo :smile:

File: server/api/routes.js [hunk 2]
```diff
@@ -18,7 +20,7 @@ router.put('/prosumer/:id/consumption', async (req, res) => {
 })
 
 // Returns the total of all prosumers consumption
-router.get('/prosumer/consumption', async (req, res) => {
+router.get('/prosumer.js/consumption', async (req, res) => {
```
Comment (edvinnn) [inline server/api/routes.js]: typo

File: server/api/routes.js [hunk 3]
```diff
@@ -33,8 +35,8 @@ router.get('/prosumer/consumption', async (req, res) => {
     }
 })
 
-// Get prosumer consumption by id
-router.get('/prosumer/:id/consumption', async (req, res) => {
+// Get prosumer.js consumption by id
+router.get('/prosumer.js/:id/consumption', async (req, res) => {
```
Comment (edvinnn) [inline server/api/routes.js]: typo

File: server/api/routes.js [hunk 4]
```diff
@@ -48,8 +50,8 @@ router.get('/prosumer/:id/consumption', async (req, res) => {
 })
 
 // PRODUCTION
-// Update prosumer production by id
-router.put('/prosumer/:id/production', async (req, res) => {
+// Update prosumer.js production by id
+router.put('/prosumer.js/:id/production', async (req, res) => {
```
Comment (edvinnn) [inline server/api/routes.js]: typo

File: server/api/routes.js [hunk 5]
```diff
@@ -63,7 +65,7 @@ router.put('/prosumer/:id/production', async (req, res) => {
 })
 
 // Returns the total of all prosumers production
-router.get('/prosumer/production', async (req, res) => {
+router.get('/prosumer.js/production', async (req, res) => {
```
Comment (edvinnn) [inline server/api/routes.js]: typo

File: server/api/routes.js [hunk 6]
```diff
@@ -78,8 +80,8 @@ router.get('/prosumer/production', async (req, res) => {
     }
 })
 
-// Get prosumer production by id
-router.get('/prosumer/:id/production', async (req, res) => {
+// Get prosumer.js production by id
+router.get('/prosumer.js/:id/production', async (req, res) => {
```
Comment (edvinnn) [inline server/api/routes.js]: typo

File: server/api/routes.js [hunk 7]
```diff
@@ -93,8 +95,8 @@ router.get('/prosumer/:id/production', async (req, res) => {
 })
 
 // NET PRODUCTION
-// Get prosumer net production by id
-router.get('/prosumer/:id/net_production', async(req, res) => {
+// Get prosumer.js net production by id
+router.get('/prosumer.js/:id/net_production', async(req, res) => {
```
Comment (edvinnn) [inline server/api/routes.js]: typo

File: server/api/routes.js [hunk 8]
```diff
@@ -111,8 +113,8 @@ router.get('/prosumer/:id/net_production', async(req, res) => {
 } )
 
 // BUFFER
-// Get prosumer buffer by id
-router.get('/prosumer/:id/buffer', async (req, res) => {
+// Get prosumer.js buffer by id
+router.get('/prosumer.js/:id/buffer', async (req, res) => {
```
Comment (edvinnn) [inline server/api/routes.js]: typo

File: server/api/routes.js [hunk 9]
```diff
@@ -125,8 +127,8 @@ router.get('/prosumer/:id/buffer', async (req, res) => {
     }
 })
 
-// Update prosumer buffer by id
-router.put('/prosumer/:id/buffer', async (req, res) => {
+// Update prosumer.js buffer by id
+router.put('/prosumer.js/:id/buffer', async (req, res) => {
```
Comment (edvinnn) [inline server/api/routes.js]: typo

File: server/api/routes.js [hunk 10]
```diff
@@ -139,8 +141,8 @@ router.put('/prosumer/:id/buffer', async (req, res) => {
     }
 })
 
-// Get prosumer max buffer by id
-router.get('/prosumer/:id/buffer_max', async (req, res) => {
+// Get prosumer.js max buffer by id
+router.get('/prosumer.js/:id/buffer_max', async (req, res) => {
```
Comment (edvinnn) [inline server/api/routes.js]: typo

File: server/api/routes.js [hunk 11]
```diff
@@ -153,8 +155,8 @@ router.get('/prosumer/:id/buffer_max', async (req, res) => {
     }
 })
 
-// Update prosumer max buffer by id
-router.put('/prosumer/:id/buffer_max', async (req, res) => {
+// Update prosumer.js max buffer by id
+router.put('/prosumer.js/:id/buffer_max', async (req, res) => {
```
Comment (edvinnn) [inline server/api/routes.js]: typo

File: server/api/routes.js [hunk 12]
```diff
@@ -167,4 +169,92 @@ router.put('/prosumer/:id/buffer_max', async (req, res) => {
...
+router.get('/price/latest', async (req, res) => {
+    try {
+        const price = await sim_db_utils.getLatestPrice()
+        res.status(200).json({price})
+    } catch (err) {
+        console.log(err)
+        res.status(500)
+    }
+})
+
```
Comment (edvinnn) [inline server/api/routes.js]: Not sure... but wasn't this implemented somewhere already? :)

File: server/api/routes.js (hunk 1 above)
Comment (Hampusholmstrom) [reply server/api/routes.js]: haha wops, im on it!

File: server/api/routes.js (hunk 12 above)
Comment (Hampusholmstrom) [reply server/api/routes.js]: Yes, they are implemented in the simulator/api/routes.js as well but I guess just for testing before with postman. simulator/api/routes.js is no longer needed right?

**Pull request merged**